### PR TITLE
fix(ext/websocket): fix closing of WebSocketStream with unread messages

### PR DESCRIPTION
**Pull request by israelidanny**
This PR fixes issue https://github.com/denoland/deno/issues/15616

The following logic is implemented:
1. When `close()` is called, we record when our close request was sent
2. If there are pending messages in the stream but the user isn't reading them (there isn't any active pull() running), we fast forward `pull()`
3. We do this until we either encounter a response close frame or if the server keeps sending other frames, a maximum of 5 seconds as per RFC 6455. The timeout is an extreme edge case in case the server is not written up to spec and does not reply with a close frame but rather ignores it and keeps sending other events.

The relevant [RFC 6455](https://www.rfc-editor.org/rfc/rfc6455#section-5.5.1) statement:
```
the client SHOULD wait for the server to
close the connection but MAY close the connection at any time after
sending and receiving a Close message, e.g., if it has not received a
TCP Close from the server in a reasonable time period.
```

✅ all tests are passing, and the original PR issue at hand is fixed when testing with a build based on this PR

## Why are there both async and sync tests? 
Due to the various race conditions that exist - those four tests behaved differently with different edge cases.

## Why the `setTimeout(() => {}, 0)`?
Due to the nature of the bug, having to do with a race condition - we wanted the event loop to queue the logic after any `pull()` that would have naturally happened. Otherwise we would be failing on one of the sync edge case tests.

### CODE REVIEW

File: ext/websocket/02_websocketstream.js [hunk 1]
```diff
@@ -370,7 +385,13 @@
         this[_earlyClose] = true;
       } else if (this[_closed].state === "pending") {
         PromisePrototypeCatch(
-          core.opAsync("op_ws_close", this[_rid], code, closeInfo.reason),
+          PromisePrototypeThen(
+            core.opAsync("op_ws_close", this[_rid], code, closeInfo.reason),
+            () => {
+              this[_closeSent] = new Date().getTime();
+              this[_pull]();
```
Comment (israelidanny) [inline ext/websocket/02_websocketstream.js:392]: this needs to be aware of `controller` as well, I'll refactor if the general direction makes sense.

**Comment by israelidanny** [comment]
@crowlKats @bartlomieju would love a review, because this fixes a real bug that we're experiencing (and it's a bug that could affect potentially any websocket client (because it triggers once you receive more messages than you're willing to read)

**Comment by crowlKats** [comment]
I am not too happy with this solution, however since there is no spec, and i am unsure what would be the right way, i will say lets go ahead with this. 
However, any pulls made, should still be enqueued to the stream

**Comment by israelidanny** [comment]
> I am not too happy with this solution, however since there is no spec, and i am unsure what would be the right way, i will say lets go ahead with this. However, any pulls made, should still be enqueued to the stream

They are, the only thing I'm doing right now is running `pull()` in a situation where it wouldn't run itself.

**Comment by israelidanny** [comment]
I see that formatter lint is failing CI, my bad. Gonna run it now.

**Comment by israelidanny** [comment]
✅ Done. All checks pass. Hope this is merge-able now 🤞 Let me know if I can do anything else @crowlKats @bartlomieju 

Also, FYI I opened a further PR on the spec explainer repo to raise the issue of defining this behaviour better at spec level:
https://github.com/ricea/websocketstream-explainer/issues/15

Once they reply I'd be more than happy to submit a second PR refining this into something a bit more elegant, but the external behaviour (these tests from this PR) will probably stay 100% the same, so it should be an easy fix.

**Comment by crowlKats** [comment]
well, the spec is basically non-existent; pretty much behaviour for everything still needs to be defined

File: ext/websocket/02_websocketstream.js [hunk 2]
```diff
@@ -370,7 +395,14 @@
...
         PromisePrototypeCatch(
-          core.opAsync("op_ws_close", this[_rid], code, closeInfo.reason),
+          PromisePrototypeThen(
+            core.opAsync("op_ws_close", this[_rid], code, closeInfo.reason),
+            () => {
+              setTimeout(() => {
+                this[_closeSent].resolve(new Date().getTime());
+              }, 0);
+            },
+          ),
```
Comment (crowlKats) [inline ext/websocket/02_websocketstream.js:405]: nitpick: dont use `PromisePrototypeCatch` & `PromisePrototypeThen`, but just `PromisePrototypeCatch` and 3rd param

Comment (israelidanny) [reply ext/websocket/02_websocketstream.js]: ✅ done

**Review by crowlKats — APPROVED** [review]
LGTM

**Pull request merged**